src/workspace.rs

```diff
@@ -0,0 +1,23 @@
+use super::error::{MainError, PnError};
+use lets_find_up::*;
+use std::path::{Path, PathBuf};
+
+const WORKSPACE_MANIFEST_FILENAME: &str = "pnpm-workspace.yaml";
+
+pub fn find_workspace_root(cwd: &dyn AsRef<Path>) -> Result<PathBuf, MainError> {
```

Comment:
Same suggestion as above,

either

```suggestion
pub fn find_workspace_root(cwd: &Path) -> Result<PathBuf, MainError> {
```

or

```suggestion
pub fn find_workspace_root(cwd: impl AsRef<Path>) -> Result<PathBuf, MainError> {
```

or

```suggestion
pub fn find_workspace_root<Cwd: AsRef<Path>>(cwd: Cwd) -> Result<PathBuf, MainError> {
```

Reply:
this is hard

Reply:
I'm sorry, I wrote a lot because I think it would be helpful for someone who is learning Rust.

Anyway, just apply the first suggestion, please.

Reply:
I know, thanks. I am just saying that it is hard 😄 I'll have to learn more about it.

Reply:
> I'll have to learn more about it.

There are some ~~rules~~ guidelines of thumbs:
* Never write `&dyn AsRef<Type>`. Instead, use `&Type` or `impl AsRef<Type>`.
* Never write `&dyn Deref<Target = Type>` or `impl Deref<Target = Type>`. Instead, just write `&Type`. There are exceptions with generic trait bounds however.

Reply:
Agree with the above. Opaque types (`impl AsRef<Type>`) generally allows code to be more portable while also avoiding double dispatch (`&dyn AsRef<Type>`).